src/terrain/terrain.js

```diff
@@ -827,15 +827,30 @@ export class Terrain extends Elevation {
     }
 
     _shouldDisableRenderCache(): boolean {
-        // Disable render caches on dynamic events due to fading.
-        const isCrossFading = id => {
+        if (!this.renderCached) {
+            return true;
+        }
+
+        // Disable render caches on dynamic events due to fading or transitioning.
+        if (this._style.light && this._style.light.hasTransition()) {
+            return true;
+        }
+
+        for (const id in this._style._sourceCaches) {
+            if (this._style._sourceCaches[id].hasTransition()) {
+                return true;
+            }
+        }
+
+        const fadingOrTransitioning = id => {
             const layer = this._style._layers[id];
-            const isHidden = !layer.isHidden(this.painter.transform.zoom);
```

Comment:
I noticed a possible bug where `isHidden` was negated twice leading to wrong result. Hidden layers are now properly skipped when determining whether caching should be disabled. @astojilj can you confirm this?

Reply:
Great find. This relates to raster fade on terrain and good to recheck the details (https://github.com/mapbox/mapbox-gl-js/commit/a192ebbd6b1d51180576b8283dcafa3e2c70c778) - there is render cache there that seems to be passing now, too.